### PR TITLE
WIP: Add a Heartbeat object to the API

### PR DESCRIPTION
**Pull request by gmarek**
Ref. #14733

@bgrant0607 PTAL. I purposefully didn't add this anywhere else, as those will be mechanical changes.

cc @wojtek-t

### CODE REVIEW

File: pkg/api/types.go [hunk 1]
```diff
@@ -1368,6 +1368,13 @@ type NodeSpec struct {
 	Unschedulable bool `json:"unschedulable,omitempty"`
 }
 
+type Heartbeat struct {
+	// Time when this heartbeat was generated.
```
Comment (wojtek-t) [inline pkg/api/types.go]: I think that you also need ObjectMeta - similarly to what we have in binding:
https://github.com/gmarek/kubernetes/blob/heartbeat_api/pkg/api/types.go#L1568

Comment (gmarek) [inline pkg/api/types.go]: I don't think I need. There's a lot of various stuff, that's completely unimportant for a hearbeat (name, uuid, etc.). Binding is a different case, as it represents a binary relation, and we store one dimension as ObjectMeta, and second as ObjectReference. Here we need only one reference - I don't really care if it's ObjectMeta, or ObjectReference, but ObjectReference is smaller, and have everything we need (or at least everything I believe we need).

Comment (bgrant0607) [inline pkg/api/types.go]: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md

File: pkg/api/types.go [hunk 2]
```diff
@@ -1368,6 +1368,13 @@ type NodeSpec struct {
...
+	Timestamp unversioned.Time `json:"timestamp,omitempty"`
```
Comment (bgrant0607) [inline pkg/api/types.go]: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#naming-conventions

**Comment by bgrant0607** [comment]
Let's discuss on the issue.

**Comment by ncdc** [comment]
@kubernetes/rh-cluster-infra

**Comment by eparis** [comment]
@k8s-bot test this [triggering because no jenkins unit test present and shippable not good enough any more]

File: pkg/api/types.go (hunk 1 above)
Comment (gmarek) [inline pkg/api/types.go]: I still think it's unnecessary, as I don't think that heartbeat would be something that "represent a persistent entity" or anything that defines "object entity". I might have missed something, as it's a long read.

File: pkg/api/types.go (hunk 2 above)
Comment (gmarek) [inline pkg/api/types.go]: What is the reason for 'no *stamp' rule? Seems pretty arbitrary to me.

File: pkg/api/types.go [hunk 3]
```diff
@@ -1353,14 +1353,14 @@ type EndpointsList struct {
 
 // NodeSpec describes the attributes that a node is created with.
 type NodeSpec struct {
-	// PodCIDR represents the pod IP range assigned to the node
+	// PodCIDR represents the pod IP range assigned to the node.
```
Comment (pmorie) [inline pkg/api/types.go]: nit: peel these off into another PR?  They're orthogonal to the point of this PR and easy to review.

**Comment by eparis** [comment]
@k8s-bot test this issue: #IGNORE

Tests have been pending for 24 hours

**Comment by bgrant0607** [comment]
Closing. Can reopen when we decide what we want/need to do.

**Pull request closed without merging**